### ISSUE TITLE
Inactive tracks should not be stopped

### ISSUE THREAD

**Issue by k0nserv** [pion/webrtc#2226]
### Your environment.
  - Version: 3.1.0
  - Browser: Any
  - Other Information - N/A
  - 
### What did you do?

1. Create a track on the browser side
2. Negotiate the track with Pion
3. Remove the track on the browser side with `removeTrack`
4. Negotiate with Pion

### What did you expect?

For Pion to change the direction of the transceiver in question to `inactive` and not stop it.

### What happened?

Pion changes the direction of the transceiver in question to `inactive` **and** stops it:

https://github.com/pion/webrtc/blob/ea7964c12c1a1e51fe9d86c8cc9cae8cbca22e58/peerconnection.go#L1058-L1061

This relates to the issue https://github.com/webrtc-rs/webrtc/issues/192 which in turns is related to https://github.com/webrtc-rs/webrtc/issues/191(I'm not sure if this happens in Pion too)

**Comment by adriancable** [comment]
@k0nserv - if it's correct that changing the direction to inactive on the Pion side shouldn't stop the track (and this sounds quite plausible, but I haven't checked the RFC) then I think removing L1058-1061 in `webrtc/peerconnection.go` is necessary but not sufficient.

We would also need to introduce the concept, I suppose, of a 'paused' track, whereby received RTP (but not RTCP?) data is discarded, and sending RTP (but not RTCP?) fails. We'd also need some way of aborting an in-progress ReadRTP and WriteRTP in that track if it enters the pause state while in the middle of those calls. I don't quite have the right mental model yet of the cleanest way to do that.

Thoughts from some experts on this? @davidzhao maybe?

**Comment by davidzhao** [comment]
Far from being an expert on this, but I think the current behavior is actually correct.

[RTPTransceiver.Stop()](https://github.com/pion/webrtc/blob/ea7964c12c1a1e51fe9d86c8cc9cae8cbca22e58/rtptransceiver.go#L151) in Pion doesn't work exactly the same as the browser. The current implementation of Stop is more like a paused state. According to [the specs](https://datatracker.ietf.org/doc/html/rfc8829#section-4.2.1), once a transceiver is paused, it cannot be restarted. In Pion, it'd happily re-use the same transceiver in order to send other tracks.

**Comment by k0nserv** [comment]
I posted a [suggestion](https://github.com/webrtc-rs/webrtc/issues/192#issuecomment-1129997236) for a paused state, for the sender, in the webrtc-rs issue. I've not got a sophisticated test bed for Pion, but in webrtc-rs if `m=` section becomes inactive it can never transition back to something else as far as I can see. Does that apply to Pion too?

> [RTPTransceiver.Stop()](https://github.com/pion/webrtc/blob/ea7964c12c1a1e51fe9d86c8cc9cae8cbca22e58/rtptransceiver.go#L151) in Pion doesn't work exactly the same as the browser. The current implementation of Stop is more like a paused state. According to [the specs](https://datatracker.ietf.org/doc/html/rfc8829#section-4.2.1), once a transceiver is paused, it cannot be restarted. In Pion, it'd happily re-use the same transceiver in order to send other tracks.

For example, does this mean you've experienced the transition from `inactive` back to e.g. `sendrecv`?

I also noticed that the `Stop` method doesn't actually update the `stopped` bool value, in fact I don't see any code path that touch that value. The [endless negotiation issue in webrtc-rs](https://github.com/webrtc-rs/webrtc/issues/191) stems from this stopping behaviour.

It's also worth noting that the doc comment for `Stop` says

> Stop **irreversibly** stops the RTPTransceiver

(Emphasis mine)

**Comment by davidzhao** [comment]
@k0nserv I believe you are right. I got a bit confused because the behavior is different depending on if it's sending or receiving.

When sending, Pion handles transceiver re-use correctly. [RemoveTrack](https://github.com/pion/webrtc/blob/ea7964c12c1a1e51fe9d86c8cc9cae8cbca22e58/peerconnection.go#L1731) puts the transceiver into an inactive state, until another Track is attached to it.

When receiving, Pion would stop the transceiver if a remote offerer has set the transceiver to inactive. Once stopped, a transceiver cannot be reused again (the comment is correct, even though we seemed to missed setting stopped flag). I *think* all we need to do here is to remove that Stop and instead set transceiver direction to inactive. [PeerConnection.configureRTPReceivers](https://github.com/pion/webrtc/blob/master/peerconnection.go#L1255) should stop and reset the receiver.

**Comment by adriancable** [comment]
@davidzhao

> I think all we need to do here is to remove that Stop and instead set transceiver direction to inactive.

Which `Stop` are you referring to? Is it the same as the one @k0nserv originally referred to? https://github.com/pion/webrtc/blob/ea7964c12c1a1e51fe9d86c8cc9cae8cbca22e58/peerconnection.go#L1058-L1061

This only gets called if `direction == RTPTransceiverDirectionInactive` already, so can you clarify what you mean when you say 'instead' set transceiver direction to inactive.

In summary, I think my question is: do you think removing L1058-L1061 is a necessary and sufficient change to fix this?

**Comment by k0nserv** [comment]
> In summary, I think my question is: do you think removing L1058-L1061 is a necessary and sufficient change to fix this?

I tested this in `webrtc-rs´, but it doesn't cause the sender to start sending again when the transceiver in question transitions from `inactive` to `sendonly` or `sendrecv´. Likely, there's some more work that's needed

**Comment by davidzhao** [comment]
> Which `Stop` are you referring to? Is it the same as the one @k0nserv originally referred to?
Yes, I'm referring to the same Stop


> https://github.com/pion/webrtc/blob/ea7964c12c1a1e51fe9d86c8cc9cae8cbca22e58/peerconnection.go#L1058-L1061
> 
> This only gets called if `direction == RTPTransceiverDirectionInactive` already, so can you clarify what you mean when you say 'instead' set transceiver direction to inactive.

That's the direction as specified in the offer, not what the local Transceiver indicates. I'm saying in this case, we need to match the local Transceiver's direction to inactive as well. It'll then trigger a receiver reset in `configureRTPReceivers`

> In summary, I think my question is: do you think removing L1058-L1061 is a necessary and sufficient change to fix this?

**Comment by adriancable** [comment]
@davidzhao - so something like this?

1. Remove L1058-1061 in `webrtc/peerconnection.go`
2. At L1104 insert:

```
			case direction == RTPTransceiverDirectionInactive:
				t.setDirection(RTPTransceiverDirectionInactive)
```

**Comment by davidzhao** [comment]
Yeah, I think that should be sufficient.. but this chain of events is sort of complex.. wdyt @Sean-Der ?

**Comment by k0nserv** [comment]
I've been [experimenting](https://github.com/webrtc-rs/webrtc/commit/986e8a7e6ece7dc812511b3bafd9eea974ea6899) with this in `webrtc-rs` and while these changes do correct the direction when moving from inactive and back the receiver ends up being stopped anyway. This happens in [`start_rtp`](https://github.com/webrtc-rs/webrtc/blob/986e8a7e6ece7dc812511b3bafd9eea974ea6899/src/peer_connection/peer_connection_internal.rs#L138). The reason for this is that `received_needs_stopped` ends up being `true` and the receiver is then [stopped](https://github.com/webrtc-rs/webrtc/blob/986e8a7e6ece7dc812511b3bafd9eea974ea6899/src/peer_connection/peer_connection_internal.rs#L193-L196).

However, this might be different for Pion due to the changes from https://github.com/pion/webrtc/pull/2087.


Another question on my mind is: How should it be detected when a transceiver transitions from .e.g. `recvonly` to `inactive`? In an SFU case this even must lead to downstream consequence for all other tracks that are sending this data. Is firing the `mute` event on the track appropriate?

@Sean-Der Do you have peer-to-peer playgrounds for testing with Chrome that you could share to explore this behaviour?

**Comment by k0nserv** [comment]
I've been playing around a bit with this in Chrome using the following playgrounds:

* [Sender](https://jsfiddle.net/mc0xvpuL/35/)
* [Receiver](https://jsfiddle.net/q72Ldcv3/22/)

When the transceiver transitions from `recvonly` in the Receiver to `inactive` it does fire `mute` on the track in question as I speculated above. However, going back from inactive to `sendonly` or `sendrecv` is not working. On the Sender side explicitly setting `sendonly` or `sendrecv` causes the offer that we then create to have a `recvonly` direction for some reason. I don't know why that this is, but perhaps it's because I don't create the transceiver manually.

**Comment by davidzhao** [comment]
> Another question on my mind is: How should it be detected when a transceiver transitions from .e.g. `recvonly` to `inactive`? In an SFU case this event must lead to downstream consequence for all other tracks that are sending this data. Is firing the `mute` event on the track appropriate?

In Pion, when `configureRTPReceivers` is called, it will stop the receiver and cause EOF to be sent to consumer of the track. I don't know if we need to fire any other events that way. `mute` is fired when there isn't data coming through, but should not be fired when the sender has negotiated the transceiver to inactive.

**Comment by davidzhao** [comment]
> The spec says that RTCP should continue to be sent but RTP media should stop.

You are right. though I'm not sure what type of RTCP it would still make sense to send. The receiving end would have marked this track as completed. When the sender wishes to send another track, it would trigger the receiver to fire `onTrack` again.

> I'm also curious about how you handle this in LiveKit, i.e. when a camera track is removed how is that handled?

Within LiveKit, the behavior differs depending on the platform. When a receiving browser sees a transceiver go from recvonly to inactive, the `MediaStreamTrack` is removed from `MediaStream`. we [handle it here](https://github.com/livekit/client-sdk-js/blob/80407b942e7fab210a8cc305378741140f191343/src/room/track/RemoteTrack.ts#L36).

On mobile platforms, it would notify us that RTPReceiver had been removed. [here's](https://github.com/livekit/client-sdk-swift/blob/main/Sources/LiveKit/Core/Transport.swift#L283) how we handle it.

The behavior on mobile seem to indicate that the RTPReceiver should be removed/stopped entirely.

**Comment by k0nserv** [comment]
Hmm, right I suppose the receiver can be removed and re-created entirely instead of entering some paused state and then resuming. However, since the specification says that RTCP should continue to flow that seems to indicate that it's decoupled from the RTCPReceiver in libWebrtc then.

**Comment by k0nserv** [comment]
FWIW there's this [test](https://webrtc.googlesource.com/src/+/refs/heads/main/pc/peer_connection_jsep_unittest.cc#583) in libWebrtc that asserts that the transceivers are not stopped in the case of an m= section becoming inactive.

**Comment by k0nserv** [comment]
Details from the spec in https://w3c.github.io/webrtc-pc/#set-the-session-description in step 7.9.1.7.2. Implementation in [libWebRTC](https://webrtc.googlesource.com/src/+/refs/heads/main/pc/sdp_offer_answer.cc#1581)

**Comment by langhuihui** [comment]
so how can I do to receive data on the server side when inactive back to sendrecv on the browser side ?
